internals/overlord/logstate/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,381 @@
+	resourceAttributes map[string][]keyValue
+}
+
+func NewClient(options *ClientOptions) *Client {
+	options = fillDefaultOptions(options)
+	c := &Client{
+		options:            options,
+		httpClient:         &http.Client{Timeout: options.RequestTimeout},
+		buffer:             make([]entryWithService, 2*options.MaxRequestEntries),
+		resourceAttributes: make(map[string][]keyValue),
+	}
+	// c.entries should be backed by the same array as c.buffer.
+	c.entries = c.buffer[:0]
+	return c
+}
+
+// ClientOptions allows overriding default parameters (e.g. for testing).
+type ClientOptions struct {
+	RequestTimeout    time.Duration
+	MaxRequestEntries int
+	UserAgent         string
+	TargetName        string
+	Location          string
+}
+
+func fillDefaultOptions(options *ClientOptions) *ClientOptions {
+	if options.RequestTimeout == 0 {
+		options.RequestTimeout = requestTimeout
+	}
+	if options.MaxRequestEntries == 0 {
+		options.MaxRequestEntries = maxRequestEntries
+	}
+	return options
+}
+
+// SetLabels sets resource attributes for a service. Labels are analogous to OpenTelemetry's resource attributes.
+func (c *Client) SetLabels(serviceName string, attributes map[string]string) {
+	if attributes == nil {
+		delete(c.resourceAttributes, serviceName)
+		return
+	}
+
+	// Convert attributes to keyValue format.
+	keyValuePairs := make([]keyValue, 0, len(attributes)+1)
+
+	// Add service.name attribute.
+	keyValuePairs = append(keyValuePairs, keyValue{
+		Key:   "service.name",
+		Value: anyValue{StringValue: &serviceName},
+	})
+
+	// Sort other labels to ensure deterministic order.
+	keys := make([]string, 0, len(attributes))
+	for k := range attributes {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		v := attributes[k]
+		keyValuePairs = append(keyValuePairs, keyValue{
+			Key:   k,
+			Value: anyValue{StringValue: &v},
+		})
+	}
+
+	c.resourceAttributes[serviceName] = keyValuePairs
+}
+
+func (c *Client) Add(entry servicelog.Entry) error {
+	if n := len(c.entries); n >= c.options.MaxRequestEntries {
+		// "entries" is full - remove the first element to make room.
+		// Zero the removed element to allow garbage collection.
+		c.entries[0] = entryWithService{}
+		c.entries = c.entries[1:]
+	}
+
+	if len(c.entries) >= cap(c.entries) {
+		// Copy all the elements to the start of the buffer.
+		copy(c.buffer, c.entries)
+
+		// Reset the view into the buffer.
+		c.entries = c.buffer[:len(c.entries):len(c.buffer)]
+
+		// Zero removed elements to allow garbage collection.
+		for i := len(c.entries); i < len(c.buffer); i++ {
+			c.buffer[i] = entryWithService{}
+		}
+	}
+
+	c.entries = append(c.entries, entryWithService{
+		entry:   encodeEntry(entry),
+		service: entry.Service,
+	})
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) logRecord {
+	message := strings.TrimSuffix(entry.Message, "\n")
+
+	return logRecord{
+		TimeUnixNano: fmt.Sprintf("%d", entry.Time.UnixNano()),
```

Comment:
It's a bit more efficient to use the more specific `strconv.FormatInt(entry.Time.UnixNano(), 10)` function that the generalized `fmt.Sprintf`.